### ISSUE TITLE
Euclidean vs Cosine distance in same-domain, cross-domain and multi-source

### ISSUE THREAD

**Issue by mikel-brostrom** [KaiyangZhou/deep-person-reid#502]
Hi @KaiyangZhou, thanks again for you great work,

I see that most of the models under: https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO use euclidean distances. Specially under same-domain ReiD. However, under cross domain I see both. And for multi-source I only see cosine. For cross-domain and multi-source, cosine seem to outperform euclidean. What is the rationale behind this?

**Comment by KaiyangZhou** [comment]
features learned with the cross-entropy loss tend to cluster like beams shooting out from the same origin (see this https://github.com/KaiyangZhou/pytorch-center-loss#results and refer to the center loss paper)

the dot product between image features and label vectors (i.e. the classification weights) essensially measures their similarity in the cosine space

therefore, using cosine distance makes more sense and often performs better (ignore the different metrics used in the model zoo because I used the euclidean distance at the early stage but switched to the cosine distance later)

**Comment by mikel-brostrom** [comment]
> ignore the different metrics used in the model zoo because I used the euclidean distance at the early stage but switched to the cosine distance later

I see. I updated the distance metric in my repo to `euclidean` based on what you wrote there but never had time to do enough experimentation to know whether this was giving better consistent results or not.

> features learned with the cross-entropy loss tend to cluster like beams shooting out from the same origin

Get it! Cosine makes more sense then.
